README.md
# particle-wave
[![npm package](https://img.shields.io/npm/v/particle-wave.svg)](https://www.npmjs.com/package/particle-wave)
### 示例
> [demo展示](https://jekorx.github.io/particle-wave)

![particle-wave](screenshot/pic1.png)
### 用法
```bash
# 安装依赖
yarn add particle-wave
# or
npm i particle-wave -S
```
> 使用
```css
/* 示例样式 */
#particle-wave {
  width: 1024px;
  height: 512px;
  background-color: #1E1D46;
}
```
```html
<!-- 示例html -->
<div id="particle-wave"></div>
```
```javascript
/**
 * 示例javascript
 */
import ParticleWave from 'particle-wave'

// 小圆点尺寸
const pointSize = 4

const pw = new ParticleWave(document.getElementById('particle-wave'), {
  uniforms: {
    size: { type: 'float', value: pointSize },
    field: { type: 'vec3', value: [0, 0, 0] },
    speed: { type: 'float', value: 7 }
  },
  onResize (w, h, dpi) {
    const position = []
    const color = []
    const width = 400 * (w / h)
    const depth = 500
    const height = 7
    const distance = 9
    for (let x = 0; x < width; x += distance) {
      for (let z = 0; z < depth; z += distance) {
        position.push(-width / 2 + x, -30, -depth / 2 + z)
        color.push(0, 1 - (x / width) * 1, 0.5 + x / width * 0.5, z / depth)
      }
    }
    if (this.uniforms) {
      this.uniforms.field = [width, height, depth]
      this.uniforms.size = (h / 400) * pointSize * dpi
    }
    if (this.buffers) {
      this.buffers.position = position
      this.buffers.color = color
    }
  }
})

document.getElementById('clear').onclick = function () {
  console.log('ParticleWave will clear', pw)
  pw.clear()
}
```


CHANGELOG.md
# 更新日志

#### 1.1.3 (2021-6-22)

* index.d.ts

#### 1.1.2 (2021-6-21)

* ts add clear method

#### 1.1.0 (2021-6-21)

* 增加typescript支持

#### 1.0.6 (2020-9-16)

* 增加clear方法


src/index.js
export default class {
  constructor (holder, options = {}) {
    options = Object.assign({
      antialias: false,
      depthTest: false,
      mousemove: false,
      autosize: true,
      side: 'front',
      vertex: `
        #define M_PI 3.14159
        precision highp float;
        attribute vec4 a_position;
        attribute vec4 a_color;
        uniform vec2 u_resolution;
        uniform vec2 u_mousemove;
        precision highp float;
        uniform float u_time;
        uniform float u_size;
        uniform float u_speed;
        uniform vec3 u_field;
        uniform mat4 u_projection;
        varying vec4 v_color;
        void main() {
          vec3 pos = a_position.xyz;
          pos.y += (
            cos(pos.x / u_field.x * M_PI * 8.0 + u_time * u_speed) +
            sin(pos.z / u_field.z * M_PI * 8.0 + u_time * u_speed)
        ) * u_field.y;
          gl_Position = u_projection * vec4(pos.xyz, a_position.w);
          gl_PointSize = (u_size / gl_Position.w) * 100.0;
          v_color = a_color;
        }`,
      fragment: `
        precision highp float;
        uniform sampler2D u_texture;
        uniform int u_hasTexture;
        varying vec4 v_color;
        void main() {
          gl_FragColor = v_color * texture2D(u_texture, gl_PointCoord);
        }`,
      uniforms: {},
      buffers: {},
      camera: {},
      texture: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAMAAABEpIrGAAAAb1BMVEUAAAD///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////8v0wLRAAAAJHRSTlMAC/goGvDhmwcExrVjWzrm29TRqqSKenRXVklANSIUE8mRkGpv+HOfAAABCElEQVQ4y4VT13LDMAwLrUHteO+R9f/fWMfO6dLaPeKVEECRxOULWsEGpS9nULDwia2Y+ALqUNbAWeg775zv+sA4/FFRMxt8U2FZFCVWjR/YrH4/H9sarclSKdPMWKzb8VsEeHB3m0shkhVCyNzeXeAQ9Xl4opEieX2QCGnwGbj6GMyjw9t1K0fK9YZunPXeAGsfJtYjwzxaBnozGGorYz0ypK2HzQSYx1y8DgSRo2ewOiyh2QWOEk1Y9OrQV0a8TiBM1a8eMHWYnRMy7CZ4t1CmyRkhSUvP3gRXyHOCLBxNoC3IJv//ZrJ/kxxUHPUB+6jJZZHrpg6GOjnqaOmzp4NDR48OLxn/H27SRQ08S0ZJAAAAAElFTkSuQmCC',
      onUpdate: () => { },
      onResize: () => { }
    }, options)

    if (!holder) return false

    const uniforms = Object.assign({
      time: { type: 'float', value: 0 },
      hasTexture: { type: 'int', value: 0 },
      resolution: { type: 'vec2', value: [0, 0] },
      mousemove: { type: 'vec2', value: [0, 0] },
      projection: { type: 'mat4', value: [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1] }
    }, options.uniforms)

    const buffers = Object.assign({
      position: { size: 3, data: [] },
      color: { size: 4, data: [] }
    }, options.buffers)

    const camera = Object.assign({
      fov: 60,
      near: 1,
      far: 1000,
      aspect: 1,
      z: 110,
      perspective: true
    }, options.camera)

    const canvas = document.createElement('canvas')
    const gl = canvas.getContext('webgl', { antialias: options.antialias })

    if (!gl) return false

    this.count = 0
    this.gl = gl
    this.canvas = canvas
    this.camera = camera
    this.holder = holder
    this.onUpdate = options.onUpdate
    this.onResize = options.onResize
    this.data = {}

    holder.appendChild(canvas)

    this.createProgram(options.vertex, options.fragment)

    this.createBuffers(buffers)
    this.createUniforms(uniforms)

    this.updateBuffers()
    this.updateUniforms()

    this.createTexture(options.texture)

    gl.enable(gl.BLEND)
    gl.enable(gl.CULL_FACE)
    gl.blendFunc(gl.SRC_ALPHA, gl.ONE)
    gl[options.depthTest ? 'enable' : 'disable'](gl.DEPTH_TEST)

    if (options.autosize) window.addEventListener('resize', e => this.resize(e), false)
    if (options.mousemove) window.addEventListener('mousemove', e => this.mousemove(e), false)

    this.resize()

    this.update = this.update.bind(this)
    this.time = { start: performance.now(), old: performance.now() }
    this.update()
  }

  mousemove (e) {
    let x = e.pageX / this.width * 2 - 1
    let y = e.pageY / this.height * 2 - 1

    this.uniforms.mousemove = [x, y]
  }

  resize (e) {
    const holder = this.holder
    const canvas = this.canvas
    const gl = this.gl

    if (!gl || !canvas || !holder) return

    const width = this.width = holder.offsetWidth
    const height = this.height = holder.offsetHeight
    const aspect = this.aspect = width / height
    const dpi = this.dpi = devicePixelRatio

    canvas.width = width * dpi
    canvas.height = height * dpi
    canvas.style.width = width + 'px'
    canvas.style.height = height + 'px'

    gl.viewport(0, 0, width * dpi, height * dpi)
    gl.clearColor(0, 0, 0, 0)

    this.uniforms.resolution = [width, height]
    this.uniforms.projection = this.setProjection(aspect)

    this.onResize(width, height, dpi)
  }

  setProjection (aspect) {
    const camera = this.camera

    if (camera.perspective) {
      camera.aspect = aspect

      const fovRad = camera.fov * (Math.PI / 180)
      const f = Math.tan(Math.PI * 0.5 - 0.5 * fovRad)
      const rangeInv = 1.0 / (camera.near - camera.far)

      const matrix = [
        f / camera.aspect, 0, 0, 0,
        0, f, 0, 0,
        0, 0, (camera.near + camera.far) * rangeInv, -1,
        0, 0, camera.near * camera.far * rangeInv * 2, 0
      ]

      matrix[14] += camera.z
      matrix[15] += camera.z

      return matrix
    } else {
      return [
        2 / this.width, 0, 0, 0,
        0, -2 / this.height, 0, 0,
        0, 0, 1, 0,
        -1, 1, 0, 1
      ]
    }
  }

  createShader (type, source) {
    const gl = this.gl
    const shader = gl.createShader(type)

    gl.shaderSource(shader, source)
    gl.compileShader(shader)
    if (gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
      return shader
    } else {
      console.log(gl.getShaderInfoLog(shader))
      gl.deleteShader(shader)
    }
  }

  createProgram (vertex, fragment) {
    const gl = this.gl

    const vertexShader = this.createShader(gl.VERTEX_SHADER, vertex)
    const fragmentShader = this.createShader(gl.FRAGMENT_SHADER, fragment)

    const program = gl.createProgram()

    gl.attachShader(program, vertexShader)
    gl.attachShader(program, fragmentShader)
    gl.linkProgram(program)

    if (gl.getProgramParameter(program, gl.LINK_STATUS)) {
      gl.useProgram(program)
      this.program = program
    } else {
      console.log(gl.getProgramInfoLog(program))
      gl.deleteProgram(program)
    }
  }

  createUniforms (data) {
    const gl = this.gl
    const uniforms = this.data.uniforms = data
    const values = this.uniforms = {}

    Object.keys(uniforms).forEach(name => {
      const uniform = uniforms[name]

      uniform.location = gl.getUniformLocation(this.program, 'u_' + name)

      Object.defineProperty(values, name, {
        set: value => {
          uniforms[name].value = value
          this.setUniform(name, value)
        },
        get: () => uniforms[name].value
      })
    })
  }

  setUniform (name, value) {
    const gl = this.gl
    const uniform = this.data.uniforms[name]

    uniform.value = value

    switch (uniform.type) {
      case 'int': {
        gl.uniform1i(uniform.location, value)
        break
      }
      case 'float': {
        gl.uniform1f(uniform.location, value)
        break
      }
      case 'vec2': {
        gl.uniform2f(uniform.location, ...value)
        break
      }
      case 'vec3': {
        gl.uniform3f(uniform.location, ...value)
        break
      }
      case 'vec4': {
        gl.uniform4f(uniform.location, ...value)
        break
      }
      case 'mat2': {
        gl.uniformMatrix2fv(uniform.location, false, value)
        break
      }
      case 'mat3': {
        gl.uniformMatrix3fv(uniform.location, false, value)
        break
      }
      case 'mat4': {
        gl.uniformMatrix4fv(uniform.location, false, value)
        break
      }
    }
  }

  updateUniforms () {
    const uniforms = this.data.uniforms
    Object.keys(uniforms).forEach(name => {
      const uniform = uniforms[name]
      this.uniforms[name] = uniform.value
    })
  }

  createBuffers (data) {
    const buffers = this.data.buffers = data
    const values = this.buffers = {}

    Object.keys(buffers).forEach(name => {
      const buffer = buffers[name]

      buffer.buffer = this.createBuffer('a_' + name, buffer.size)

      Object.defineProperty(values, name, {
        set: data => {
          buffers[name].data = data
          this.setBuffer(name, data)

          if (name === 'position') this.count = buffers.position.data.length / 3
        },
        get: () => buffers[name].data
      })
    })
  }

  createBuffer (name, size) {
    const gl = this.gl
    const program = this.program

    const index = gl.getAttribLocation(program, name)
    const buffer = gl.createBuffer()

    gl.bindBuffer(gl.ARRAY_BUFFER, buffer)
    gl.enableVertexAttribArray(index)
    gl.vertexAttribPointer(index, size, gl.FLOAT, false, 0, 0)

    return buffer
  }

  setBuffer (name, data) {
    const gl = this.gl
    const buffers = this.data.buffers

    if (name == null && !gl.bindBuffer(gl.ARRAY_BUFFER, null)) return

    gl.bindBuffer(gl.ARRAY_BUFFER, buffers[name].buffer)
    gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(data), gl.STATIC_DRAW)
  }

  updateBuffers () {
    const buffers = this.buffers
    Object.keys(buffers).forEach(name => {
      buffers[name] = name.data
    })

    this.setBuffer(null)
  }

  createTexture (src) {
    const gl = this.gl
    const texture = gl.createTexture()

    gl.bindTexture(gl.TEXTURE_2D, texture)
    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, 1, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE, new Uint8Array([0, 0, 0, 0]))

    this.texture = texture

    if (src) {
      this.uniforms.hasTexture = 1
      this.loadTexture(src)
    }
  }

  loadTexture (src) {
    const gl = this.gl
    const texture = this.texture

    const textureImage = new Image()

    textureImage.onload = () => {
      gl.bindTexture(gl.TEXTURE_2D, texture)

      gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, textureImage)

      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR)
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR)

      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE)
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE)
    }
    textureImage.src = src
  }

  update () {
    const gl = this.gl
    if (!gl) return

    const now = performance.now()
    const elapsed = (now - this.time.start) / 5000
    const delta = now - this.time.old
    this.time.old = now

    this.uniforms.time = elapsed

    if (this.count > 0) {
      gl.clear(gl.COLOR_BUFFER_BIT)
      gl.drawArrays(gl.POINTS, 0, this.count)
    }

    this.onUpdate(delta)

    requestAnimationFrame(this.update)
  }

  clear () {
    const gl = this.gl
    const holder = this.holder
    if (!gl || !holder) return
    gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT | gl.STENCIL_BUFFER_BIT)
    gl.getAttachedShaders(this.program).forEach(shader => {
      gl.deleteShader(shader)
    })
    gl.deleteProgram(this.program)
    Object.values(this.data.buffers).forEach(({ buffer }) => {
      gl.deleteBuffer(buffer)
    })
    gl.deleteTexture(this.texture)
    this.program = null
    this.data = null
    this.buffers = null
    this.texture = null
    this.camera = null
    this.holder = null
    this.canvas = null
    this.gl = null
  }

}


types/index.d.ts
declare module 'particle-wave'

export default class ParticleWave {
  constructor(el: HTMLElement | null, options: {
    antialias?: boolean,
    depthTest?: boolean,
    mousemove?: boolean,
    autosize?: boolean,
    side?: string,
    vertex?: string,
    fragment?: string,
    uniforms?: {
      time?: {
        type: string,
        value: number
      },
      hasTexture?: {
        type: string,
        value: number
      },
      speed?: {
        type: string,
        value: number
      },
      resolution?: {
        type: string,
        value: number[]
      },
      mousemove?: {
        type: string,
        value: number[]
      },
      projection?: {
        type: string,
        value: number[]
      },
      field?: {
        type: string,
        value: number[]
      } | number[],
      size?: {
        type: string,
        value: number
      } | number
    },
    buffers?: {
      position: number[],
      color: number[]
    },
    texture?: string,
    onUpdate?: (delta: number) => void,
    onResize?: (width: number, height: number, dpi: number) => void
  })

  clear: () => void
}


lib/particle-wave.js
module.exports=function(e){var t={};function r(n){if(t[n])return t[n].exports;var a=t[n]={i:n,l:!1,exports:{}};return e[n].call(a.exports,a,a.exports,r),a.l=!0,a.exports}return r.m=e,r.c=t,r.d=function(e,t,n){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},r.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(r.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var a in e)r.d(n,a,function(t){return e[t]}.bind(null,a));return n},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="",r(r.s=0)}([function(e,t,r){"use strict";function n(e){return function(e){if(Array.isArray(e))return a(e)}(e)||function(e){if("undefined"!=typeof Symbol&&null!=e[Symbol.iterator]||null!=e["@@iterator"])return Array.from(e)}(e)||function(e,t){if(e){if("string"==typeof e)return a(e,t);var r=Object.prototype.toString.call(e).slice(8,-1);return"Map"===(r="Object"===r&&e.constructor?e.constructor.name:r)||"Set"===r?Array.from(e):"Arguments"===r||/^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(r)?a(e,t):void 0}}(e)||function(){throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.")}()}function a(e,t){(null==t||t>e.length)&&(t=e.length);for(var r=0,n=new Array(t);r<t;r++)n[r]=e[r];return n}function i(e,t){for(var r=0;r<t.length;r++){var n=t[r];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}r.r(t),r.d(t,"default",(function(){return o}));var o=function(){function e(t){var r=this,n=1<arguments.length&&void 0!==arguments[1]?arguments[1]:{};if(function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")}(this,e),n=Object.assign({antialias:!1,depthTest:!1,mousemove:!1,autosize:!0,side:"front",vertex:"\n        #define M_PI 3.14159\n        precision highp float;\n        attribute vec4 a_position;\n        attribute vec4 a_color;\n        uniform vec2 u_resolution;\n        uniform vec2 u_mousemove;\n        precision highp float;\n        uniform float u_time;\n        uniform float u_size;\n        uniform float u_speed;\n        uniform vec3 u_field;\n        uniform mat4 u_projection;\n        varying vec4 v_color;\n        void main() {\n          vec3 pos = a_position.xyz;\n          pos.y += (\n            cos(pos.x / u_field.x * M_PI * 8.0 + u_time * u_speed) +\n            sin(pos.z / u_field.z * M_PI * 8.0 + u_time * u_speed)\n        ) * u_field.y;\n          gl_Position = u_projection * vec4(pos.xyz, a_position.w);\n          gl_PointSize = (u_size / gl_Position.w) * 100.0;\n          v_color = a_color;\n        }",fragment:"\n        precision highp float;\n        uniform sampler2D u_texture;\n        uniform int u_hasTexture;\n        varying vec4 v_color;\n        void main() {\n          gl_FragColor = v_color * texture2D(u_texture, gl_PointCoord);\n        }",uniforms:{},buffers:{},camera:{},texture:"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAMAAABEpIrGAAAAb1BMVEUAAAD///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////8v0wLRAAAAJHRSTlMAC/goGvDhmwcExrVjWzrm29TRqqSKenRXVklANSIUE8mRkGpv+HOfAAABCElEQVQ4y4VT13LDMAwLrUHteO+R9f/fWMfO6dLaPeKVEECRxOULWsEGpS9nULDwia2Y+ALqUNbAWeg775zv+sA4/FFRMxt8U2FZFCVWjR/YrH4/H9sarclSKdPMWKzb8VsEeHB3m0shkhVCyNzeXeAQ9Xl4opEieX2QCGnwGbj6GMyjw9t1K0fK9YZunPXeAGsfJtYjwzxaBnozGGorYz0ypK2HzQSYx1y8DgSRo2ewOiyh2QWOEk1Y9OrQV0a8TiBM1a8eMHWYnRMy7CZ4t1CmyRkhSUvP3gRXyHOCLBxNoC3IJv//ZrJ/kxxUHPUB+6jJZZHrpg6GOjnqaOmzp4NDR48OLxn/H27SRQ08S0ZJAAAAAElFTkSuQmCC",onUpdate:function(){},onResize:function(){}},n),!t)return!1;var a=Object.assign({time:{type:"float",value:0},hasTexture:{type:"int",value:0},resolution:{type:"vec2",value:[0,0]},mousemove:{type:"vec2",value:[0,0]},projection:{type:"mat4",value:[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]}},n.uniforms),i=Object.assign({position:{size:3,data:[]},color:{size:4,data:[]}},n.buffers),o=Object.assign({fov:60,near:1,far:1e3,aspect:1,z:110,perspective:!0},n.camera),u=document.createElement("canvas"),s=u.getContext("webgl",{antialias:n.antialias});if(!s)return!1;this.count=0,this.gl=s,this.canvas=u,this.camera=o,this.holder=t,this.onUpdate=n.onUpdate,this.onResize=n.onResize,this.data={},t.appendChild(u),this.createProgram(n.vertex,n.fragment),this.createBuffers(i),this.createUniforms(a),this.updateBuffers(),this.updateUniforms(),this.createTexture(n.texture),s.enable(s.BLEND),s.enable(s.CULL_FACE),s.blendFunc(s.SRC_ALPHA,s.ONE),s[n.depthTest?"enable":"disable"](s.DEPTH_TEST),n.autosize&&window.addEventListener("resize",(function(e){return r.resize(e)}),!1),n.mousemove&&window.addEventListener("mousemove",(function(e){return r.mousemove(e)}),!1),this.resize(),this.update=this.update.bind(this),this.time={start:performance.now(),old:performance.now()},this.update()}var t,r;return t=e,(r=[{key:"mousemove",value:function(e){var t=e.pageX/this.width*2-1;e=e.pageY/this.height*2-1;this.uniforms.mousemove=[t,e]}},{key:"resize",value:function(e){var t,r,n,a=this.holder,i=this.canvas,o=this.gl;o&&i&&a&&(t=this.width=a.offsetWidth,r=this.height=a.offsetHeight,n=this.aspect=t/r,a=this.dpi=devicePixelRatio,i.width=t*a,i.height=r*a,i.style.width=t+"px",i.style.height=r+"px",o.viewport(0,0,t*a,r*a),o.clearColor(0,0,0,0),this.uniforms.resolution=[t,r],this.uniforms.projection=this.setProjection(n),this.onResize(t,r,a))}},{key:"setProjection",value:function(e){var t=this.camera;if(t.perspective){t.aspect=e;var r=t.fov*(Math.PI/180);e=Math.tan(.5*Math.PI-.5*r),r=1/(t.near-t.far);return(r=[e/t.aspect,0,0,0,0,e,0,0,0,0,(t.near+t.far)*r,-1,0,0,t.near*t.far*r*2,0])[14]+=t.z,r[15]+=t.z,r}return[2/this.width,0,0,0,0,-2/this.height,0,0,0,0,1,0,-1,1,0,1]}},{key:"createShader",value:function(e,t){var r=this.gl;e=r.createShader(e);if(r.shaderSource(e,t),r.compileShader(e),r.getShaderParameter(e,r.COMPILE_STATUS))return e;console.log(r.getShaderInfoLog(e)),r.deleteShader(e)}},{key:"createProgram",value:function(e,t){var r=this.gl,n=this.createShader(r.VERTEX_SHADER,e);e=this.createShader(r.FRAGMENT_SHADER,t),t=r.createProgram();r.attachShader(t,n),r.attachShader(t,e),r.linkProgram(t),r.getProgramParameter(t,r.LINK_STATUS)?(r.useProgram(t),this.program=t):(console.log(r.getProgramInfoLog(t)),r.deleteProgram(t))}},{key:"createUniforms",value:function(e){var t=this,r=this.gl,n=this.data.uniforms=e,a=this.uniforms={};Object.keys(n).forEach((function(e){n[e].location=r.getUniformLocation(t.program,"u_"+e),Object.defineProperty(a,e,{set:function(r){n[e].value=r,t.setUniform(e,r)},get:function(){return n[e].value}})}))}},{key:"setUniform",value:function(e,t){var r=this.gl,a=this.data.uniforms[e];switch(a.value=t,a.type){case"int":r.uniform1i(a.location,t);break;case"float":r.uniform1f(a.location,t);break;case"vec2":r.uniform2f.apply(r,[a.location].concat(n(t)));break;case"vec3":r.uniform3f.apply(r,[a.location].concat(n(t)));break;case"vec4":r.uniform4f.apply(r,[a.location].concat(n(t)));break;case"mat2":r.uniformMatrix2fv(a.location,!1,t);break;case"mat3":r.uniformMatrix3fv(a.location,!1,t);break;case"mat4":r.uniformMatrix4fv(a.location,!1,t)}}},{key:"updateUniforms",value:function(){var e=this,t=this.data.uniforms;Object.keys(t).forEach((function(r){var n=t[r];e.uniforms[r]=n.value}))}},{key:"createBuffers",value:function(e){var t=this,r=this.data.buffers=e,n=this.buffers={};Object.keys(r).forEach((function(e){var a=r[e];a.buffer=t.createBuffer("a_"+e,a.size),Object.defineProperty(n,e,{set:function(n){r[e].data=n,t.setBuffer(e,n),"position"===e&&(t.count=r.position.data.length/3)},get:function(){return r[e].data}})}))}},{key:"createBuffer",value:function(e,t){var r=this.gl,n=this.program;n=r.getAttribLocation(n,e),e=r.createBuffer();return r.bindBuffer(r.ARRAY_BUFFER,e),r.enableVertexAttribArray(n),r.vertexAttribPointer(n,t,r.FLOAT,!1,0,0),e}},{key:"setBuffer",value:function(e,t){var r=this.gl,n=this.data.buffers;null==e&&!r.bindBuffer(r.ARRAY_BUFFER,null)||(r.bindBuffer(r.ARRAY_BUFFER,n[e].buffer),r.bufferData(r.ARRAY_BUFFER,new Float32Array(t),r.STATIC_DRAW))}},{key:"updateBuffers",value:function(){var e=this.buffers;Object.keys(e).forEach((function(t){e[t]=t.data})),this.setBuffer(null)}},{key:"createTexture",value:function(e){var t=this.gl,r=t.createTexture();t.bindTexture(t.TEXTURE_2D,r),t.texImage2D(t.TEXTURE_2D,0,t.RGBA,1,1,0,t.RGBA,t.UNSIGNED_BYTE,new Uint8Array([0,0,0,0])),this.texture=r,e&&(this.uniforms.hasTexture=1,this.loadTexture(e))}},{key:"loadTexture",value:function(e){var t=this.gl,r=this.texture,n=new Image;n.onload=function(){t.bindTexture(t.TEXTURE_2D,r),t.texImage2D(t.TEXTURE_2D,0,t.RGBA,t.RGBA,t.UNSIGNED_BYTE,n),t.texParameteri(t.TEXTURE_2D,t.TEXTURE_MIN_FILTER,t.LINEAR),t.texParameteri(t.TEXTURE_2D,t.TEXTURE_MAG_FILTER,t.LINEAR),t.texParameteri(t.TEXTURE_2D,t.TEXTURE_WRAP_S,t.CLAMP_TO_EDGE),t.texParameteri(t.TEXTURE_2D,t.TEXTURE_WRAP_T,t.CLAMP_TO_EDGE)},n.src=e}},{key:"update",value:function(){var e,t,r,n=this.gl;n&&(t=((e=performance.now())-this.time.start)/5e3,r=e-this.time.old,this.time.old=e,this.uniforms.time=t,0<this.count&&(n.clear(n.COLOR_BUFFER_BIT),n.drawArrays(n.POINTS,0,this.count)),this.onUpdate(r),requestAnimationFrame(this.update))}},{key:"clear",value:function(){var e=this.gl,t=this.holder;e&&t&&(e.clear(e.COLOR_BUFFER_BIT|e.DEPTH_BUFFER_BIT|e.STENCIL_BUFFER_BIT),e.getAttachedShaders(this.program).forEach((function(t){e.deleteShader(t)})),e.deleteProgram(this.program),Object.values(this.data.buffers).forEach((function(t){t=t.buffer,e.deleteBuffer(t)})),e.deleteTexture(this.texture),this.program=null,this.data=null,this.buffers=null,this.texture=null,this.camera=null,this.holder=null,this.canvas=null,this.gl=null)}}])&&i(t.prototype,r),e}()}]).default;

src/demo.js
import ParticleWave from './index.js'

const pointSize = 4

const pw = new ParticleWave(document.getElementById('app'), {
  uniforms: {
    size: { type: 'float', value: pointSize },
    field: { type: 'vec3', value: [0, 0, 0] },
    speed: { type: 'float', value: 7 }
  },
  onResize(w, h, dpi) {
    const position = []
    const color = []
    const width = 400 * (w / h)
    const depth = 500
    const height = 7
    const distance = 9
    for (let x = 0; x < width; x += distance) {
      for (let z = 0; z < depth; z += distance) {
        position.push(-width / 2 + x, -30, -depth / 2 + z)
        color.push(0, 1 - (x / width) * 1, 0.5 + x / width * 0.5, z / depth)
      }
    }
    this.uniforms.field = [width, height, depth]
    this.buffers.position = position
    this.buffers.color = color
    this.uniforms.size = (h / 400) * pointSize * dpi
  }
})
document.getElementById('clear').onclick = function () {
  console.log('ParticleWave will clear', pw)
  pw.clear()
}
